Add csg boolean operators using elalish/manifold.

fire opened the pull request:
Fixes: https://github.com/godotengine/godot-proposals/issues/9711

- [x] Need to discuss with @elalish how to trim the dependent libraries
- [x] Need to discuss with @elalish about removing exceptions
- [x] Add copyright notices.
- [x] Pending work to remove Thrust
- [x] Pending work to remove exceptions

![image](https://github.com/godotengine/godot/assets/32321/e41057de-510d-45a5-b8ad-34be2e345e18)

https://github.com/godotengine/godot/assets/32321/28312b80-ce1d-4293-8203-22de021cbe39

fire commented on modules/csg/csg_shape.cpp:
```diff
@@ -162,6 +165,151 @@ void CSGShape3D::_make_dirty(bool p_parent_removing) {
 	dirty = true;
 }
 
+enum {
```
Some somewhere enums, but not left bare here.

fire committed:
Add csg boolean operators using elalish/manifold.

fire commented on modules/csg/csg_shape.cpp:
```diff
@@ -30,7 +30,317 @@
 
+#include "core/math/aabb.h"
 #include "core/math/geometry_2d.h"
+#include "core/math/math_funcs.h"
```
Not all of these are used.

AThousandShips replied:
All the `core/math` ones are already included, as is `vector` and `ref_counted`

elalish commented on modules/csg/SCsub:
```diff
@@ -6,6 +6,64 @@ Import("env_modules")
+        "src/manifold/src/subdivision.cpp",
+        "src/manifold/src/csg_tree.cpp",
+        "src/cross_section/src/cross_section.cpp",
+        "src/third_party/quickhull/QuickHull.cpp",
```
This is good practice for me, since I haven't seen a large project integrate our library this way before. I would like to make it easier if possible, and more modular. I take it you don't want to integrate us as a submodule? 

If you're not interested in 2D tools or convex hulls, it should be possible to drop those dependencies, but we could probably structure our library better for doing this. @pca006132 do you know any way to allow e.g. building Manifold without quickhull, so that the `Manifold::ConvexHull` methods are just stubs?

pca006132 replied:
Some kind of `#if` will work, change them to throwing exceptions when user don't want to compile quickhull.

fire replied:
As a general rule, Godot Engine vendors its entire dependencies because it makes development much easier for contributors and for CI/CD. I am interested in the other libraries, but it would be best if I can find ways to shrink the binary size.

Although, if your quickhull is good, @lawnjelly has been in that area and can evaluate the quality.

elalish replied:
Would love to hear your evaluation - in fact we have a GSoC contributor who just started and is looking into improving our hulling - either using a different dependency or writing a new one. https://github.com/opencax/GSoC/issues/89

fire replied:
To be honest my best experience is with https://github.com/SarahWeiii/CoACD for convex hulling.

elalish commented:
Manifold should already be set up to work just fine without exceptions if you don't compile with them - is it giving you any trouble?

pca006132 commented:
> Manifold should already be set up to work just fine without exceptions if you don't compile with them - is it giving you any trouble?

Probably not, I forgot to only throw exceptions for vector out of range when user enable compiling with exceptions. And tbb uses exceptions iirc.

fire commented on modules/csg/csg_shape.cpp:
```diff
@@ -32,6 +32,306 @@
+		}
+	}
+
+	void pack_manifold(const float p_snap, HashMap<uint32_t, Ref<Material>> &r_materials) {
```
@elalish Any recommendations to improve this? I got stuck translating your tutorial.

elalish replied:
Did you see my review below? Does it make sense?

fire replied:
I wasn't able to understand your tutorial, it's probably something on my part.

elalish commented on modules/csg/csg_shape.cpp:
```diff
@@ -32,6 +32,306 @@
+				current_brush.faces.push_back(face);
+				unique_face_i++;
+			}
+			manifold::MeshGL mesh;
```
The problem is here: you're creating a mesh per material, but what you want is a single mesh where each material becomes a new run of triangles. Basically you need to be concatenating what you're doing here instead of starting over. Then perform the `Merge` outside of your materials loop.

31 replied:
I'm going to try to take a look (chatted with @fire, I think it's mainly a matter of not having time to dig into this). I poked around and I believe the tutorial/example is [bindings/wasm/examples/three.ts](https://github.com/elalish/manifold/blob/c01f5df49a9fd5a766cfde55cba5412dc7d77f5e/bindings/wasm/examples/three.ts#L118-L144), but let me know if there's something more involved you both were using. I think I broadly understand the suggestion, though.

elalish replied:
Great to hear, and yes, you're correct. Don't hesitate to reach out if you have questions.

31 replied:
I think I got it working: https://github.com/31/godot/commit/df33cbff3b4a68941debc1c9b2c8fe9b3ba7d97d. I don't have a great set of test cases to go through, but it seems to work and keep track of materials. I rewrote a chunk of the pack/unpack functionality while I was trying to understand it better, apologies @fire for making the diff a little noisy.

31 committed:
Add convex hull CSG node

Support a material, add an icon, add simple documentation.

Co-Authored-By: 31 <31eee384@gmail.com>

31 committed:
Track materials using runs during pack and unpack

Manifold supports tracking materials associated with a run of triangles.
This commit changes Godot's Manifold pack and unpack logic to use that
functionality rather than storing the material id in each vertex.

fire committed:
Revert "Add convex hull CSG node"

This reverts commit 58516f743ec7bcbfde4429a83d0735be81202804.

fire committed:
Do not force LONG_JMP to wasm mode.

31 commented:
* Filed https://github.com/godotengine/godot/issues/93240
  * Some screenshots in here of a situation that Manifold handles *way* better than the existing meshing, even when there's an abnormality due to Snap being too high. The scene is also quite expensive but it's easy to pass around because it relies on a generated mesh rather than assets--could be useful to look at performance. (I'm very impressed, but I was looking at 4.2.2 vs. a dev build and didn't use a stopwatch, so not apples to apples.)

fire commented:
Superseded by: https://github.com/godotengine/godot/pull/94321.
